board.go

```diff
@@ -0,0 +1,141 @@
+package main
+
+import (
+	"errors"
+	"fmt"
+)
+
+// Orientation is a type that indicates in which direction the piece is moved.
+type Orientation int
+
+// Location is a type that represents a 2 dimensional coordinate.
+type Location struct {
+	x int
+	y int
+}
+
+// Direction constants.
+const (
+	Up Orientation = iota
+	Right
+	Down
+	Left
+)
+
+// Board is a structure that represents the puzzle board.
+type Board struct {
+	// Width contains the width of the puzzle board.
+	Width int
+
+	// Height contains the height of the puzzle board.
+	Height int
+
+	slots []*Piece
+
+	pieces map[*Piece]Location
+}
+
+// NewBoard constructs a new empty puzzle board.
+func NewBoard(width, height int) *Board {
+	return &Board{
+		Width:  width,
+		Height: height,
+		slots:  make([]*Piece, width*height),
+		pieces: make(map[*Piece]Location),
+	}
+}
+
+// AddPiece adds a Piece instance at the specified location.  If the piece
+// would overlap with an existing piece, it is not placed and an error is
+// returned.
+func (p *Board) AddPiece(piece *Piece, x, y int) error {
+	for i := 0; i < piece.Width; i++ {
+		for j := 0; j < piece.Height; j++ {
+			if index, err := p.GetIndex(x+i, y+j); err != nil {
+				return err
+			} else if p.slots[index] != nil {
+				return fmt.Errorf("Piece overlaps at %d, %d\n", x+i, y+j)
+			}
+		}
+	}
+
+	p.pieces[piece] = Location{
+		x: x,
+		y: y,
+	}
+
+	for i := 0; i < piece.Width; i++ {
+		for j := 0; j < piece.Height; j++ {
+			index, _ := p.GetIndex(x+i, y+j)
+			p.slots[index] = piece
+		}
+	}
+
+	return nil
+}
+
+// GetIndex calculates the index of a specific location within the pieces
+// slice.
+func (p *Board) GetIndex(x, y int) (int, error) {
```

Comment:
I'm not really a fan of those, since it's not really a 2D slice, it's more of a slice of slices.   That way each subslice *can* be a different size.  Yuk!